text/0000-custom-dst.md

```diff
@@ -0,0 +1,479 @@
+
+As of right now, the lack of custom DSTs in Rust means that we can't communicate
+with C in important ways - we lack the ability to define a `CStr` in the
+language such that `&CStr` is compatible with `char const *`,
+and we lack the ability to communicate nicely with C code that uses
+[Flexible Array Members](https://en.wikipedia.org/wiki/Flexible_array_member).
+This RFC attempts to fix this,
+as well as introduce more correctness to existing practices.
+
+Apart from FFI, it also has usecases for indexing and slicing 2-d arrays.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There's a new language trait in the standard library, similar to `Sized`,
+under `std::marker`:
+
+```rust
+unsafe trait DynamicallySized {
+    /*
+        note: these are all required due to trait implementations for
+          - Unpin - all pointer types
+          - Copy + Send + Sync - &T, &mut T
+          - Eq + Ord - *const T, *mut T
+    */
+    type Metadata: 'static + Copy + Send + Sync + Eq + Ord + Unpin;
+
+    fn size_of_val(&self) -> usize;
+    fn align_of_val(meta: Self::Metadata) -> usize;
+}
+```
+
+with an automatic implementation for all `Sized` types:
+
+```rust
+// note: this is _only_ for explanation
+// this should happen in the compiler
+unsafe impl<T> DynamicallySized for T {
+    type Metadata = ();
+
+    fn size_of_val(&self) -> usize { size_of::<T>() }
+    fn align_of_val((): ()) -> usize { align_of::<T>() }
+}
+```
+
+If you have a type which you would like to be unsized,
+you can implement this trait for your type!
+
+```rust
+#[repr(C)]
+struct CStr([c_char; 0]);
+
+unsafe impl DynamicallySized for CStr {
+    type Metadata = ();
+
+    fn size_of_val(&self) -> usize { strlen(&self.0 as *const c_char) + 1 }
+    fn align_of_val((): ()) -> usize { 1 }
+}
+```
+
+and your type will be `!Sized`.
+
+The existing `DynamicallySized` types will continue to work;
+if one writes a `DynamicallySized` type `T`,
+and then wraps `T` into a struct, they'll get the obvious semantics.
+
+```rust
+struct Foo {
+    x: usize,
+    y: CStr,
+}
+
+// size_of_val(&foo) returns size_of_header::<Foo>() + size_of_val(&foo.y)
+// same with align_of_val - simply `align_of_header::<Foo>()`
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+By language trait, we mean that `DynamicallySized` is a lang item.
+
+In addition to the explanation given above,
+we will also introduce three functions into the standard library,
+in `core::raw`, which allow you to create and destructure these
+pointers to `DynamicallySized` types:
+
+```rust
+mod core::raw {
+    pub fn from_raw_parts<T: ?Sized>(
+        ptr: *const (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *const T;
+
+    pub fn from_raw_parts_mut<T: ?Sized>(
+        ptr: *mut (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *mut T;
+
+    pub fn metadata<T: ?Sized>(
+        ptr: *const T,
```

Comment:
What happens if `ptr` is dangling?

Reply:
I'm not sure what you mean - how does `metadata` get the `<T as DynamicallySized>::Metadata` type to return? Your `Plane` example contains a call to `std::raw::metadata(self)` - what would happen if a null pointer was passed instead of `self`?

Reply:
When `T` is a dynamically-sized type, `*const T` is a "fat pointer". An instance of `*const T` can be thought of as being similar to an instance of `struct FatPtr<T> { address: usize, metadata: T::Metadata }`. ([For example, see the definition for `*const [T]` in current Rust.](https://doc.rust-lang.org/src/core/slice/mod.rs.html#63-67)) If the pointer is null, the metadata still exists (this would be like having an instance `FatPtr { address: 0, metadata: something }`). The `metadata` function doesn't need to read the value located at `address` to get the metadata; the metadata is part of the fat pointer itself.